src/ff/prime_field.rs

```diff
@@ -0,0 +1,307 @@
+
+impl BitAnd for Fp2 {
+    type Output = Self;
+
+    fn bitand(self, rhs: Self) -> Self::Output {
+        Self((self.0 & rhs.0) & 1)
+    }
+}
+
+impl BitAndAssign for Fp2 {
+    fn bitand_assign(&mut self, rhs: Self) {
+        *self = *self & rhs;
+    }
+}
+
+impl BitOr for Fp2 {
+    type Output = Self;
+
+    fn bitor(self, rhs: Self) -> Self::Output {
+        Self((self.0 | rhs.0) & 1)
+    }
+}
+
+impl BitOrAssign for Fp2 {
+    fn bitor_assign(&mut self, rhs: Self) {
+        *self = *self | rhs;
+    }
+}
+
+impl BitXor for Fp2 {
+    type Output = Self;
+
+    fn bitxor(self, rhs: Self) -> Self::Output {
+        Self((self.0 ^ rhs.0) & 1)
+    }
+}
+
+impl BitXorAssign for Fp2 {
+    fn bitxor_assign(&mut self, rhs: Self) {
+        *self = *self ^ rhs;
+    }
+}
+
+impl Not for Fp2 {
+    type Output = Self;
+
+    fn not(self) -> Self::Output {
+        Self((self.0 + 1) & 1)
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub struct Fp31(<Self as Field>::Integer);
+
+impl Field for Fp31 {
+    type Integer = u8;
+    const PRIME: Self::Integer = 31;
+    const ZERO: Self = Fp31(0);
+    const ONE: Self = Fp31(1);
+}
+
+field_impl! { Fp2, Fp31 }
+
+#[cfg(test)]
+mod test {
+    use super::{Field, Fp2, Fp31};
+    use std::ops::Mul;
+
+    #[test]
+    fn fp2() {
+        let x = Fp2::from(false);
+        let y = Fp2::from(true);
+
+        assert_eq!(Fp2(1), x + y);
+        assert_eq!(Fp2(0), x * y);
+        assert_eq!(Fp2(1), x - y);
+
+        let mut x = Fp2(1);
+        x += Fp2(1);
+        assert_eq!(Fp2(0), x);
+    }
+
+    #[test]
+    fn fp2_binary_op() {
+        let zero = Fp2::ZERO;
+        let one = Fp2::ONE;
+
+        assert_eq!(one, one & one);
+        assert_eq!(zero, zero & one);
+        assert_eq!(zero, one & zero);
+        assert_eq!(zero, zero & zero);
+        assert_eq!(zero, Fp2(31) & Fp2(32));
+        assert_eq!(one, Fp2(31) & Fp2(63));
+
+        assert_eq!(zero, zero | zero);
+        assert_eq!(one, one | one);
+        assert_eq!(one, zero | one);
+        assert_eq!(one, one | zero);
+        assert_eq!(one, Fp2(31) | Fp2(32));
+        assert_eq!(zero, Fp2(32) | Fp2(64));
+
+        assert_eq!(zero, zero ^ zero);
+        assert_eq!(one, zero ^ one);
+        assert_eq!(one, one ^ zero);
+        assert_eq!(zero, one ^ one);
+        assert_eq!(one, Fp2(31) ^ Fp2(32));
+        assert_eq!(zero, Fp2(32) ^ Fp2(64));
+
+        assert_eq!(one, !zero);
+        assert_eq!(zero, !one);
+        assert_eq!(one, !Fp2(32));
+        assert_eq!(zero, !Fp2(31));
+    }
+
+    #[test]
+    fn fp31() {
+        let x = Fp31(24);
+        let y = Fp31(23);
+
+        assert_eq!(Fp31(16), x + y);
+        assert_eq!(Fp31(25), x * y);
+        assert_eq!(Fp31(1), x - y);
+
+        let mut x = Fp31(1);
+        x += Fp31(2);
+        assert_eq!(Fp31(3), x);
+    }
+
+    #[test]
+    fn zero() {
+        macro_rules! gen_zero_test {
```

Comment:
Don't use a macro for this, use a generic function.  And invoke that function from two `#[test]` functions.